[filters] Refatoring VoxelGridCovariance to make it multi-thread safe (and more)

koide3 opened the pull request:
This PR updates VoxelGridCovariance so that it can be used in multi-threaded NDT https://github.com/PointCloudLibrary/pcl/pull/4135

- getNeighborhoodAtPoint is generalized so that any neighborhood patterns can be designated
- const qualifiers are added to getNeighborhoodAtPoint, nearestKSearch, and radiusSearch to clarify that they are thread-safe

Although the second modification slightly changes the behavior of the methods in case a voxel found by ```kdtree_``` doesn't exist in ```leaves_```, this condition never happens because ```kdtree_``` is built from voxel coordinates in ```leaves_```.

Regarding https://github.com/PointCloudLibrary/pcl/pull/4180#discussion_r449546738:
I empirically found that using adjacent + center voxels (7 voxels) is the best for accuracy. I updated the doc comment of getAdjacentVoxelsAtPoint to clarify that it retrieves 7 voxels.

SergioRAgostinho commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -361,14 +361,43 @@ namespace pcl
 
+      /** \brief Get the voxels surrounding point p designated by #relative_coordinates.
+       * \note Only voxels containing a sufficient number of points are used.
+       * \param[in] relative_coordinates relative coordinates of neighboring voxels
```
Can you provide some intuition on how this variable needs to be structured. I see it represent some notion of adjacency, represented by indices, but I'm clueless on how to supply inputs to this.

koide3 replied:
I refined the comment to provide more information on the relative coordinate matrix structure.

SergioRAgostinho commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -425,7 +454,13 @@ namespace pcl
-          k_leaves.push_back (&leaves_[voxel_centroids_leaf_indices_[k_index]]);
+          auto voxel = leaves_.find(voxel_centroids_leaf_indices_[k_index]);
+          if (voxel == leaves_.end()) {
+            PCL_WARN("Accessing to non-existent voxel");
```
I would remove this warning. The policy for showing warnings is limited to cases where you break a user's expectation somehow and is really important that he is aware of it.

That being said and before doing anything here, how realistic is it to trigger this condition. Is this supposed to ever happen. I see you're using indices supplied by a kdtree search.

koide3 replied:
I removed the warning.

I think this condition never happens unless ```kdtree_``` is modified from outside of the class, because ```kdtree_``` is built from ```leaves_```, and thus every point in ```kdtree_``` should have a corresponding voxel in ```leaves_```. 

https://github.com/PointCloudLibrary/pcl/blob/6aaade34381fbb608208172e623f71719565fd33/filters/include/pcl/filters/voxel_grid_covariance.h#L274
https://github.com/PointCloudLibrary/pcl/blob/6aaade34381fbb608208172e623f71719565fd33/filters/include/pcl/filters/voxel_grid_covariance.h#L269
https://github.com/PointCloudLibrary/pcl/blob/6aaade34381fbb608208172e623f71719565fd33/filters/include/pcl/filters/impl/voxel_grid_covariance.hpp#L306

SergioRAgostinho commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -479,7 +514,13 @@ namespace pcl
-          k_leaves.push_back (&leaves_[voxel_centroids_leaf_indices_[k_index]]);
+          auto voxel = leaves_.find(voxel_centroids_leaf_indices_[k_index]);
+          if(voxel == leaves_.end()) {
+            PCL_WARN("Accessing to non-existent voxel");
```
same comment here.

koide3 replied:
The warning removed.

koide3 commented on filters/include/pcl/filters/impl/voxel_grid_covariance.hpp:
```diff
@@ -371,28 +371,25 @@ pcl::VoxelGridCovariance<PointT>::applyFilter (PointCloud &output)
     if ((diff2min <= displacement.array ()).all () && (diff2max >= displacement.array ()).all ())
     {
-      typename std::map<std::size_t, Leaf>::iterator leaf_iter = leaves_.find (((ijk + displacement - min_b_).dot (divb_mul_)));
+      typename std::map<std::size_t, Leaf>::const_iterator leaf_iter = leaves_.find (((ijk + displacement - min_b_).dot (divb_mul_)));
```
Thanks for reviewing. I replaced const_iterator with const auto.

koide3 commented on filters/include/pcl/filters/impl/voxel_grid_covariance.hpp:
```diff
@@ -371,28 +371,25 @@ pcl::VoxelGridCovariance<PointT>::applyFilter (PointCloud &output)
 //////////////////////////////////////////////////////////////////////////////////////////
 template<typename PointT> int
-pcl::VoxelGridCovariance<PointT>::getNeighborhoodAtPoint (const PointT& reference_point, std::vector<LeafConstPtr> &neighbors)
+pcl::VoxelGridCovariance<PointT>::getNeighborhoodAtPoint (const Eigen::MatrixXi& relative_coordinates, const PointT& reference_point, std::vector<LeafConstPtr> &neighbors) const
```
I replaced ```MatrixXi``` with ```Matrix<int, 3, Dynamic>``` to clarify that it's 3D coordinates.

Although 27 (3x3x3) would be the maximum size of the matrix in practice, someone may want to use more neighbors (e.g., 5x5x5), and thus I think we cannot define the maximum size.

SergioRAgostinho commented on filters/include/pcl/filters/impl/voxel_grid_covariance.hpp:
```diff
@@ -401,7 +398,38 @@ pcl::VoxelGridCovariance<PointT>::getNeighborhoodAtPoint (const PointT& referenc
+template<typename PointT> int
+pcl::VoxelGridCovariance<PointT>::getAdjacentVoxelsAtPoint(const PointT& reference_point, std::vector<LeafConstPtr> &neighbors) const
+{
+  Eigen::MatrixXi relative_coordinates(3, 7);
```
Replace this with `Eigen::Matrix<int, 3, Eigen::Dynamic>`  to avoid building another temporary object (`Eigen::MatrixXi` -> `const Eigen::Matrix<int, 3, Eigen::Dynamic>&`) once you pass it as an argument to the function.

koide3 replied:
```MatrixXi``` is replaced with ```Matrix<int, 3, Dynamic>```

SergioRAgostinho commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -425,7 +454,12 @@ namespace pcl
-          k_leaves.push_back (&leaves_[voxel_centroids_leaf_indices_[k_index]]);
+          auto voxel = leaves_.find(voxel_centroids_leaf_indices_[k_index]);
+          if (voxel == leaves_.end()) {
+            continue;
```
should you call `--k` before continuing?

kunaltyagi replied:
Maybe replace `return k;` with `return k_leaves.size()`?

koide3 replied:
I fixed the code according to @kunaltyagi

SergioRAgostinho commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -479,7 +513,12 @@ namespace pcl
-          k_leaves.push_back (&leaves_[voxel_centroids_leaf_indices_[k_index]]);
+          auto voxel = leaves_.find(voxel_centroids_leaf_indices_[k_index]);
+          if(voxel == leaves_.end()) {
+            continue;
```
same question

koide3 replied:
same as the above

kunaltyagi commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -361,14 +361,43 @@ namespace pcl
+       * \note Only voxels containing a sufficient number of points are used.
+       * \param[in] reference_point the point to get the leaf structure at
+       * \param[out] neighbors
+       * \return number of neighbors found
```
It's a constant 1 and it's not the number of neighbors

koide3 replied:
It can return 0 if there is no voxel at ```p```. I updated the comment to indicate that the number of found neighbors is up to 1.

kunaltyagi commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -361,14 +361,43 @@ namespace pcl
+       * \note Only voxels containing a sufficient number of points are used.
+       * \param[in] reference_point the point to get the leaf structure at
+       * \param[out] neighbors
+       * \return number of neighbors found
```
Please mention the constant value of 7 here

koide3 replied:
Same as the above.

kunaltyagi commented on filters/include/pcl/filters/impl/voxel_grid_covariance.hpp:
```diff
@@ -371,28 +371,25 @@ pcl::VoxelGridCovariance<PointT>::applyFilter (PointCloud &output)
-  Eigen::Vector4i ijk (static_cast<int> (std::floor (reference_point.x / leaf_size_[0])), 
-                       static_cast<int> (std::floor (reference_point.y / leaf_size_[1])), 
-                       static_cast<int> (std::floor (reference_point.z / leaf_size_[2])), 0);
+  Eigen::Vector4i ijk = (reference_point.getArray4fMap() / leaf_size_.array()).template cast<int>();
```
We have `inverse_leaf_size_` for speeding up the division. See L334

koide3 replied:
I replaced ```leaf_size_``` with ```inverse_leaf_size_```

kunaltyagi commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -361,14 +361,43 @@ namespace pcl
+      int
+      getNeighborhoodAtPoint (const Eigen::Matrix<int, 3, Eigen::Dynamic>& relative_coordinates, const PointT& reference_point, std::vector<LeafConstPtr> &neighbors) const;
+
       /** \brief Get the voxels surrounding point p, not including the voxel containing point p.
```
Please state the number of neighbors considered (26?)

koide3 replied:
I updated the comment.

koide3 commented:
@kunaltyagi 
> Should the added function require Eigen<int, 4, Dynamic> instead since that's the sole use in the function?

I prefer ```Eigen<int, 3, Dynamic``` because we don't need to worry about what values should be placed at the fourth column. If it's ok, I would like to keep the current form.

> What happens if the neighborhood of a point contains the point itself? Does the accuracy increase in a similar fashion as in case of adjacent voxels?

Possibly yes. I tried to use 27 neighbors once, and it was better than 26 neighbors (in both accuracy and processing speed). But, it was much slower than 7 neighbors, and the accuracy gain was not significant. If you like, we can add a function to retrieve 27 neighbors as well.

> Do we need a better naming than nighborhood and adjacentvoxels? I can see someone getting confused between the 2

Should we rename them like ```get7NeighborhoodAtPoint``` and ```get26NeighborhoodAtPoint```?

kunaltyagi commented:
> Should we rename them like get7NeighborhoodAtPoint and get26NeighborhoodAtPoint?

`getFaceNeighbors` and `AllNeighbors`? The second makes much less sense than the former.

koide3 commented on filters/include/pcl/filters/voxel_grid_covariance.h:
```diff
@@ -479,9 +513,14 @@ namespace pcl
         for (const int &k_index : k_indices)
         {
-          k_leaves.push_back (&leaves_[voxel_centroids_leaf_indices_[k_index]]);
+          auto voxel = leaves_.find(voxel_centroids_leaf_indices_[k_index]);
```
I added const to some unchanged variables.

koide3 commented:
> getFaceNeighbors and AllNeighbors? The second makes much less sense than the former.
> That'd be nice since it's better than 26 neighbors.

I added a method to get 3x3x3 neighbors and changed the method names (7neighbors: ```getFaceNeighborsAtPoint```, 27: ```getAllNeighborsAtPoint```)

I considered to change the name of 26neighbors method to ```getSurroundingVoxelsAtPoint```, but I noticed that it's public. Is it allowed to change the public method name? (AFAIK, it's used only in NDT though)

kunaltyagi commented:
Thanks for the effort @koide3 

Would you please rebase the commits into
* New functionality related changed
* Modification and refactoring of existing code
* White space changes (if needed)

The code doesn't need to compile for all commits (but should work fine as a whole)

kunaltyagi commented:
Any changes (rebasing of commits) or we can go ahead and merge (squash and merge)?

SergioRAgostinho commented:
I was gonna squash merge. 😬

kunaltyagi commented:
You didn't so I thought you were waiting on something. Merging now.
